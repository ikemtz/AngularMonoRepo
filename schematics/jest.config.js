module.exports = {
  testMatch: ["**/+(*.)+(spec|test).+(ts)?(x)"],
  transform: {
    "^.+\\.(ts|js|html)$": "ts-jest",
  },
  testPathIgnorePatterns: ["/node_modules/", "/dist/"],
  moduleFileExtensions: ["ts", "js", "html"],
  collectCoverage: true,
  coverageReporters: ["html", "lcov", "cobertura"],
  reporters: [
    "default",
    [
      "jest-junit",
      {
        outputDirectory: ".",
        outputName: "./junit.schematics.xml",
      },
    ],
  ],
  moduleNameMapper: {},
};
