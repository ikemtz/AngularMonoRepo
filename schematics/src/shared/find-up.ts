import { dirname, parse, resolve } from "path";
import { lstatSync } from "fs";

export function findUpSync(name: string): string | undefined {
  let directory = __dirname;
  const { root } = parse(directory);

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const foundPath = locatePathSync(name, directory);
    if (foundPath) {
      return resolve(directory, foundPath);
    }
    else if (directory === root) {
      break;
    } else {
      directory = dirname(directory);
    }
  }
}

export function locatePathSync(
  file: string,
  directory: string
): string | undefined {
  try {
    const stat = lstatSync(resolve(directory, file));

    if (stat['isFile']) {
      return file;
    }
    // eslint-disable-next-line no-empty
  } catch { }
}