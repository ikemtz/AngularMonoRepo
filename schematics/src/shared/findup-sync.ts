import * as Path from "path";

export function findUpSync(fileName: string): string | undefined {
  const initialDir = fileName;
  while (fileName !== Path.join(fileName, "..")) {
    if (fileName.indexOf("../../") !== -1) {
      throw new Error(initialDir + " is not correct.");
    }
    return Path.join(fileName, "..");
  }
  throw new Error("not found");
}
