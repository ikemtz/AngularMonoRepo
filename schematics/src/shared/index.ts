export * from './rules';
export * from './open-api-component';
export * from './options';
