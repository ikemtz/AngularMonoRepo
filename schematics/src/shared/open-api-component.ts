export interface OpenApiComponent {
  properties: Properties;
  required?: string[];
}
export interface Properties {
  [key: string]: PropertyInfo;
}
export interface PropertyInfo {
  name?: string;
  enum?: [string | number];
  type: 'string' | 'boolean' | 'array' | 'integer' | 'number' | 'date';
  format?: 'date-time' | 'uuid' | 'int32' | 'double';
  nullable?: boolean;
  readOnly?: boolean;
  required: boolean;
  htmlInputType: 'text' | 'number' | 'checkbox' | 'date';
  snakeCaseName: string;
  maxLength: number;
  testFactoryValue: string;
}
