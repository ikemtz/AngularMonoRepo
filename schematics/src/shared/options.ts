import { PropertyInfo } from './open-api-component';

export interface IOptions {
  /** Entity Name (Must match schema name in json) */
  name: string;
  /** Parent Entity Name (Must match schema name in json) */
  parentName?: string;
  /** NGRX Store Feature Name */
  storeName?: string;
  /** Angular application Prefix */
  appPrefix: string;
  /** Folder to place generated files */
  path: string;
  /** Optional: Url of remote OpenApi JSON document (If this is not provided, then you must specify openApiJsonFileName) */
  openApiJsonUrl?: string;
  /** Optional: File path of local OpenApi JSON document (If this is not provided, then you must specify openApiJsonUrl) */
  openApiJsonFileName?: string;
  /** Optional: This is the property that will used for initial sorting on list and auto focus on crud schematics */
  modelFolderLocation?: string;
  /** Optional: Location of model files relative to the module/component generated root */

  firstProperty?: PropertyInfo;
  /** Ignore: This is for internal schematic use only  */
  swaggerProperties: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  swaggerObjectProperties?: PropertyInfo[];
  /** Ignore: This is for internal schematic use only  */
  hasDates?: boolean;
  hasObjects?: boolean;
  hasNullableDates?: boolean;
}
