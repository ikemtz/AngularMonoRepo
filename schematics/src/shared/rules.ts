import {
  apply,
  applyTemplates,
  chain,
  DirEntry,
  mergeWith,
  move,
  Rule,
  SchematicContext,
  Tree,
  url,
} from '@angular-devkit/schematics';
import { strings, normalize } from '@angular-devkit/core';
import fetch from 'node-fetch';
import { PropertyInfo, OpenApiComponent } from './open-api-component';
import * as pluralize from 'pluralize';
import { Observable, from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { IOptions } from './options';
import _ = require('lodash');
import { TslintFixTask } from '@angular-devkit/schematics/tasks';
import * as fs from 'fs';
import * as findUp from 'find-up';
import * as https from 'https';
import * as http from 'http';

export function getSwaggerDoc(options: IOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    let jsonDoc: Observable<any> | null = null;
    if (options.openApiJsonFileName) {
      const fileName = getFileNames(options.openApiJsonFileName);
      context.logger.info(`Getting Swagger Document @${fileName}`);
      if (!fileName) {
        throw new Error(`File not found: ${options.openApiJsonFileName}`);
      }
      const response = fs.readFileSync(fileName);
      const apiDoc = JSON.parse(response.toString());
      jsonDoc = of(apiDoc);
    } else if (options.openApiJsonUrl) {
      context.logger.info(`Getting Swagger Document @${options.openApiJsonUrl}`);
      const httpsAgent = options.openApiJsonUrl.toLowerCase().startsWith("https") ? new https.Agent({
        rejectUnauthorized: false,
      }) : new http.Agent();
      jsonDoc = from(fetch(options.openApiJsonUrl, {
        agent: httpsAgent
      })).pipe(
        switchMap(resp => from(resp.json())));
    }
    if (jsonDoc) {
      return jsonDoc.pipe(map(data => processOpenApiDoc(data, options, host)));
    }
    return of(host);

  };
}

function getFileNames(openApiJsonFileName: string) {
  if (fs.existsSync(`${__dirname}/${openApiJsonFileName}`)) {
    return `${__dirname}/${openApiJsonFileName}`;
  }
  return findUp.sync(openApiJsonFileName);
}

export function processOpenApiDoc(data: any, schema: IOptions, host: Tree) {
  const openApiComonent = data.components.schemas[strings.classify(schema.name)] as OpenApiComponent;
  if (!openApiComonent) {
    throw new Error(`OpenApi Component not found in swagger doc: ${schema.name}`);
  }
  const properties = openApiComonent.properties as {
    [key: string]: PropertyInfo;
  };
  const filteredProperties: PropertyInfo[] = [];
  const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];
  schema.hasDates = false;
  for (const propertyKey in properties) {
    if (excludedFields.indexOf(propertyKey) < 0 && properties[propertyKey].type !== 'array') {
      const property =
        mapPropertyAttributes(schema, properties[propertyKey], {
          ...properties[propertyKey],
          name: propertyKey,
          required: (openApiComonent.required ?? []).indexOf(propertyKey) > -1,
        });
      filteredProperties.push(property);
    }
  }
  schema.swaggerProperties = filteredProperties;
  return host;
}

function mapPropertyAttributes(options: IOptions, source: PropertyInfo, dest: any) {

  if (source.type === 'number' || source.type === 'integer') {
    dest.htmlInputType = 'number';
    dest.filterExpression = 'numeric';
    dest.testFactoryValue = '0';
  } else if (source.type === 'boolean') {
    dest.htmlInputType = 'checkbox';
    dest.filterExpression = 'boolean';
    dest.testFactoryValue = 'true';
  } else if (source.format === 'date-time') {
    dest.htmlInputType = 'date';
    dest.filterExpression = 'date';
    dest.testFactoryValue = 'new Date()';
    options.hasDates = true;
  } else {
    dest.htmlInputType = 'text';
    dest.filterExpression = 'text';
    dest.testFactoryValue = `'${dest.maxLength ?
      _.snakeCase(dest.name).toUpperCase().substring(0, dest.maxLength) : _.snakeCase(dest.name).toUpperCase()}'`;
  }
  dest.snakeCaseName = _.snakeCase(dest.name);
  dest.startCaseName = _.startCase(dest.name);
  return dest;
}

export function generateFiles(schema: IOptions, templateType: 'list' | 'crud' | 'module'): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.logger.info(`Running ${strings.capitalize(templateType)} Schematic`);

    const singularizedName = pluralize(schema.name, 1);
    const pluralizedName = pluralize(schema.name, 2);
    const singularizedStoreName = schema.storeName ? pluralize(schema.storeName, 1) : '';
    const pluralizedStoreName = schema.storeName ? pluralize(schema.storeName, 2) : '';

    const folderName = normalize(`${schema.path}/${strings.dasherize(pluralizedName)}-${templateType}`);
    const templateSource = apply(url('./files'), [
      applyTemplates({
        ...strings,
        ...schema,
        pluralizedName,
        singularizedName,
        pluralizedStoreName,
        singularizedStoreName,
      }),
      move(folderName),
    ]);
    return chain([mergeWith(templateSource)])(tree, context);
  };
}

export function runLint(schema: IOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const dir: DirEntry | null = tree.getDir(schema.path.substr(0, schema.path.lastIndexOf('/')));
    const files = tree.actions.reduce((acc: Set<string>, action) => {
      const path = action.path.substr(1); // Remove the starting '/'.
      if (path.endsWith('.ts') && dir && action.path.startsWith(dir.path)) {
        acc.add(path);
      }
      return acc;
    }, new Set<string>());
    context.addTask(
      new TslintFixTask({
        ignoreErrors: false,
        tsConfigPath: 'tsconfig.json',
        files: Array.from(files),
      }),
    );
    return tree;
  };
}
