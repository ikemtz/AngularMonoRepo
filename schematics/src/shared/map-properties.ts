import * as pluralize from 'pluralize';
import { PropertyInfo, OpenApiComponent, OpenApiDocument } from './open-api-component';
import { IOptions } from './options';
import { mapPropertyAttributes } from './rules';

const excludedFields = ['createdBy', 'createdOnUtc', 'tenantId', 'updatedBy', 'updatedOnUtc'];

export function mapProperties(properties: { [key: string]: PropertyInfo; }, options: IOptions, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  let firstProperty: PropertyInfo | undefined;
  const filteredProperties: PropertyInfo[] = [];
  for (const propertyKey in properties) {
    const originalProperty = properties[propertyKey];
    if (excludedFields.indexOf(propertyKey) < 0 && originalProperty.type !== 'array') {
      const property = mapPropertyAttributes(options, originalProperty, {
        ...properties[propertyKey],
        name: propertyKey,
        required: (openApiComponent.required ?? []).indexOf(propertyKey) > -1,
      });

      mapReferencedProperties(property, propertyKey, openApiComponent, openApiDoc);
      if (!firstProperty && propertyKey !== 'id') {
        firstProperty = property;
      }
      filteredProperties.push(property);
    }
  }
  return { firstProperty, filteredProperties };
}
function mapReferencedProperties(property: PropertyInfo, propertyKey: string, openApiComponent: OpenApiComponent, openApiDoc: OpenApiDocument) {
  const prop = openApiComponent.properties[propertyKey];
  const propTypeNames = prop['$ref']?.split('/');
  if (propTypeNames) {
    const typeName: string = propTypeNames.pop() || '';
    property.propertyTypeName = typeName;
    property.pluralizedPropertyTypeName = pluralize.plural(typeName);
    property.singularizedPropertyTypeName = pluralize.singular(typeName);
    const refComponent = openApiDoc.components.schemas[typeName];
    property.enum = refComponent.enum;
    if (property.enum) {
      const value = property.enum[0].toString().split(' ').pop();
      property.testFactoryValue = `${property.pluralizedPropertyTypeName}.${value}`
    }
  }
}

