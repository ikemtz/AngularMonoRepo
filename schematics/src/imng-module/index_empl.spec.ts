import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { IOptions } from '../shared';
import * as pluralize from 'pluralize';
import { readFirst } from 'imng-ngrx-utils/testing';
import { classify, dasherize } from '@angular-devkit/core/src/utils/strings';

const collectionPath = path.join(__dirname, `../collection.json`);

describe(`imng-module`, () => {
  it(`works - employees`, async () => {
    const runner = new SchematicTestRunner(`schematics`, collectionPath);
    const options: IOptions = {
      name: `employee`,
      openApiJsonUrl: `https://im-wa-empo-nrcrn.azurewebsites.net/swagger/v1/swagger.json`,
      openApiJsonFileName: '../../open-api-docs/nrcrn-empl-odata.json',
      path: `./test`,
      swaggerProperties: [],
      storeName: `employees`,
      appPrefix: ''
    };
    const tree: UnitTestTree = await readFirst(runner.runSchematicAsync(`imng-module`, options, Tree.empty()));

    expect(tree.files).toEqual([
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}.module.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}.module.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}.routing.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/+state/${dasherize(options.name)}.actions.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/+state/${dasherize(options.name)}.effects.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/+state/${dasherize(options.name)}.reducer.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/+state/${dasherize(options.name)}.selectors.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/index.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.component.html`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.component.scss`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.component.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.component.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.facade.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.facade.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/add-edit.component.html`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/add-edit.component.scss`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/add.component.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/add.component.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/api.service.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/base-entry.component.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/crud.facade.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/crud.facade.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/edit.component.spec.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/edit.component.ts`,
      `/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/index.ts`,
    ]);

    const effectsFile = tree.get(`/test/${pluralize(dasherize(options.name))}-module/+state/${dasherize(options.name)}.effects.ts`);
    let content = effectsFile?.content.toString();
    expect(content).toContain(`${options.name}ActionTypes.load${classify(pluralize(options.name))}Request(partialState[from${classify(pluralize(options.name))}Reducer.${pluralize(options.name).toUpperCase()}_FEATURE_KEY].gridODataState))),`);
    expect(content).toContain(`import { ${classify(options.name)}ApiService } from '../${dasherize(pluralize(options.name))}-crud';`);

    const listFacadeSpecFile = tree.get(`/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-list/list.facade.spec.ts`);
    content = listFacadeSpecFile?.content.toString();
    expect(content).toContain(`expect(httpClient.get).toBeCalledWith('${dasherize(pluralize(options.name))}-odata/odata/v1/${classify(pluralize(options.name))}?&$count=true');`);

    const htmlFile = tree.get(`/test/${pluralize(dasherize(options.name))}-module/${dasherize(pluralize(options.name))}-crud/add-edit.component.html`,);
    content = htmlFile?.content.toString();
    expect(content).toContain(`<input type="text" maxlength="25" class="form-control" [formControlName]="props.MOBILE_PHONE" />`);
  });
});
