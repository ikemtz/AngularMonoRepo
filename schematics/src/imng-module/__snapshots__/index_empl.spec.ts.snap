// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imng-module generation works - employees 1`] = `
[
  "/test/employees-module/+state/employee-crud.effects.ts",
  "/test/employees-module/+state/employee-list.effects.ts",
  "/test/employees-module/+state/employee.actions.ts",
  "/test/employees-module/+state/employee.feature.ts",
  "/test/employees-module/+state/index.ts",
  "/test/employees-module/employees-crud/add-edit.component.html",
  "/test/employees-module/employees-crud/add-edit.component.scss",
  "/test/employees-module/employees-crud/add.component.spec.ts",
  "/test/employees-module/employees-crud/add.component.ts",
  "/test/employees-module/employees-crud/api.service.ts",
  "/test/employees-module/employees-crud/base-entry.component.ts",
  "/test/employees-module/employees-crud/crud.facade.spec.ts",
  "/test/employees-module/employees-crud/crud.facade.ts",
  "/test/employees-module/employees-crud/edit.component.spec.ts",
  "/test/employees-module/employees-crud/edit.component.ts",
  "/test/employees-module/employees-crud/index.ts",
  "/test/employees-module/employees-list/index.ts",
  "/test/employees-module/employees-list/list.component.html",
  "/test/employees-module/employees-list/list.component.scss",
  "/test/employees-module/employees-list/list.component.spec.ts",
  "/test/employees-module/employees-list/list.component.ts",
  "/test/employees-module/employees-list/list.facade.spec.ts",
  "/test/employees-module/employees-list/list.facade.ts",
  "/test/employees-module/employees-list/list.grid-state.ts",
  "/test/employees-module/employees.module.spec.ts",
  "/test/employees-module/employees.module.ts",
  "/test/employees-module/employees.routing.ts",
]
`;
